### PR TITLE
[LoopUnroll] Add CSE to remove redundant loads after unrolling.

### PR DESCRIPTION
**Pull request by fhahn**
This patch adds loadCSE support to simplifyLoopAfterUnroll. It is based
on EarlyCSE's implementation using ScopeHashTable and is using SCEV for
accessed pointers to check to find redundant loads after unrolling.

This applies to the late unroll pass only, for full unrolling those
redundant loads will be cleaned up by the regular pipeline.

The current approach constructs MSSA on-demand per-loop, but there is
still small but notable compile-time impact:

stage1-O3  +0.04%
stage1-ReleaseThinLTO +0.06%
stage1-ReleaseLTO-g +0.05%
stage1-O0-g +0.02%
stage2-O3 +0.09%
stage2-O0-g +0.04%
stage2-clang +0.02%

https://llvm-compile-time-tracker.com/compare.php?from=c089fa5a729e217d0c0d4647656386dac1a1b135&to=ec7c0f27cb5c12b600d9adfc8543d131765ec7be&stat=instructions:u

This benefits some workloads with runtime-unrolling disabled,
where users use pragmas to force unrolling, as well as with
runtime unrolling enabled.

On SPEC/MultiSource, this removes a number of loads after unrolling
on AArch64 with runtime unrolling enabled.

```
External/S...te/526.blender_r/526.blender_r    96
MultiSourc...rks/mediabench/gsm/toast/toast    39
SingleSource/Benchmarks/Misc/ffbench            4
External/SPEC/CINT2006/403.gcc/403.gcc         18
MultiSourc.../Applications/JM/ldecod/ldecod     4
MultiSourc.../mediabench/jpeg/jpeg-6a/cjpeg     6
MultiSourc...OE-ProxyApps-C/miniGMG/miniGMG     9
MultiSourc...e/Applications/ClamAV/clamscan     4
MultiSourc.../MallocBench/espresso/espresso     3
MultiSourc...dence-flt/LinearDependence-flt     2
MultiSourc...ch/office-ispell/office-ispell     4
MultiSourc...ch/consumer-jpeg/consumer-jpeg     6
MultiSourc...ench/security-sha/security-sha    11
MultiSourc...chmarks/McCat/04-bisect/bisect     3
SingleSour...tTests/2020-01-06-coverage-009    12
MultiSourc...ench/telecomm-gsm/telecomm-gsm    39
MultiSourc...lds-flt/CrossingThresholds-flt    24
MultiSourc...dence-dbl/LinearDependence-dbl     2
External/S...C/CINT2006/445.gobmk/445.gobmk     6
MultiSourc...enchmarks/mafft/pairlocalalign    53
External/S...31.deepsjeng_r/531.deepsjeng_r     3
External/S...rate/510.parest_r/510.parest_r    58
External/S...NT2006/464.h264ref/464.h264ref    29
External/S...NT2017rate/502.gcc_r/502.gcc_r    45
External/S...C/CINT2006/456.hmmer/456.hmmer     6
External/S...te/538.imagick_r/538.imagick_r    18
External/S.../CFP2006/447.dealII/447.dealII     4
MultiSourc...OE-ProxyApps-C++/miniFE/miniFE    12
External/S...2017rate/525.x264_r/525.x264_r    36
MultiSourc...Benchmarks/7zip/7zip-benchmark    33
MultiSourc...hmarks/ASC_Sequoia/AMGmk/AMGmk     2
MultiSourc...chmarks/VersaBench/8b10b/8b10b     1
MultiSourc.../Applications/JM/lencod/lencod   116
MultiSourc...lds-dbl/CrossingThresholds-dbl    24
MultiSource/Benchmarks/McCat/05-eks/eks        15
```

### CODE REVIEW

**Comment by aeubanks** [comment]
what about conditionally running EarlyCSE if LoopUnroll made a change?

**Comment by aeubanks** [comment]
I guess that would be extending #81275 to also run EarlyCSE

**Comment by fhahn** [comment]
> what about conditionally running EarlyCSE if LoopUnroll made a change?
> I guess that would be extending #81275 to also run EarlyCSE

That would be another option, but I think we would need a loop-only version of EarlyCSE to avoid unnecessary compile-time increases. Also, regular EarlyCSE is not enough to eliminate loads that become redundant across unrolled iterations; to do that, this patch uses SCEV expressions as key to handle pointers from different iterations (which are equal due to different offsets).

Happy to go down that route if we agree in general that we want to handle such cases by running (sets of) passes depending on results from earlier passes. I think overall that would be preferable, as it makes it easier to independently test those extra passes.

**Comment by fhahn** [comment]
I prototyped this as separate patch, and the compile-time impact appears noticeably larger https://llvm-compile-time-tracker.com/compare.php?from=5c3d001668ec6117045a9750a1f9d7e3995adfee&to=12953d86d1d0a9400b0f91b9a9cc949060ab9589&stat=instructions:u

**Comment by fhahn** [comment]
ping :) 

Looks like the separate pass approach adds notably more overhead.

**Comment by fhahn** [comment]
ping :)

**Comment by fhahn** [comment]
ping, any suggestions how to move forward with this?

**Comment by fhahn** [comment]
ping

File: llvm/lib/Analysis/MemorySSA.cpp [hunk 1]
```diff
@@ -1239,7 +1240,28 @@ MemorySSA::MemorySSA(Function &Func, AliasAnalysis *AA, DominatorTree *DT)
   // make queries about all the instructions in the Function.
   assert(AA && "No alias analysis?");
   BatchAAResults BatchAA(*AA);
-  buildMemorySSA(BatchAA);
+  buildMemorySSA(BatchAA, iterator_range(F->begin(), F->end()));
+  // Intentionally leave AA to nullptr while building so we don't accidently
```
Comment (nikic) [inline llvm/lib/Analysis/MemorySSA.cpp:1244]:
```suggestion
  // Intentionally leave AA to nullptr while building so we don't accidentally
```

File: llvm/lib/Analysis/MemorySSA.cpp [hunk 2]
```diff
@@ -1537,11 +1560,26 @@ void MemorySSA::buildMemorySSA(BatchAAResults &BAA) {
   // Now do regular SSA renaming on the MemoryDef/MemoryUse. Visited will get
   // filled in with all blocks.
   SmallPtrSet<BasicBlock *, 16> Visited;
-  renamePass(DT->getRootNode(), LiveOnEntryDef.get(), Visited);
+  if (L) {
+    if (auto *P = getMemoryAccess(L->getLoopPreheader())) {
+      for (Use &U : make_early_inc_range(P->uses())) {
```
Comment (nikic) [inline llvm/lib/Analysis/MemorySSA.cpp:1565]: Braces

File: llvm/lib/Analysis/MemorySSA.cpp [hunk 3]
```diff
@@ -1537,11 +1560,26 @@ void MemorySSA::buildMemorySSA(BatchAAResults &BAA) {
   // Now do regular SSA renaming on the MemoryDef/MemoryUse. Visited will get
   // filled in with all blocks.
   SmallPtrSet<BasicBlock *, 16> Visited;
-  renamePass(DT->getRootNode(), LiveOnEntryDef.get(), Visited);
+  if (L) {
```
Comment (nikic) [inline llvm/lib/Analysis/MemorySSA.cpp:1563]: This bit of code probably deserves a comment.

File: llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp [hunk 4]
```diff
@@ -1601,6 +1605,7 @@ PreservedAnalyses LoopUnrollPass::run(Function &F,
   SmallPriorityWorklist<Loop *, 4> Worklist;
   appendLoopsToWorklist(LI, Worklist);
 
+  SmallVector<Loop *> LoopsForCSE;
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp:1608]: Dead?

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 5]
```diff
@@ -209,13 +212,143 @@ static bool isEpilogProfitable(Loop *L) {
...
+Value *getMatchingValue(LoadValue LV, LoadInst *LI, unsigned CurrentGeneration,
+                        MemorySSA *MSSA) {
+  if (!LV.DefI)
+    return nullptr;
+  if (LV.Generation != CurrentGeneration) {
+    if (!MSSA)
+      return nullptr;
+    auto *EarlierMA = MSSA->getMemoryAccess(LV.DefI);
+    MemoryAccess *LaterDef;
+    LaterDef = MSSA->getWalker()->getClobberingMemoryAccess(LI);
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/LoopUnroll.cpp:265]: Combine with declaration.

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 6]
```diff
@@ -209,13 +212,143 @@ static bool isEpilogProfitable(Loop *L) {
...
+        if (!Load || !Load->isSimple()) {
+          if (I.mayWriteToMemory())
+            CurrentGeneration++;
+          continue;
+        }
+
+        const SCEV *PtrSCEV = SE.getSCEV(Load->getPointerOperand());
+        LoadValue LV = AvailableLoads.lookup(PtrSCEV);
+        if (Value *M =
+                getMatchingValue(LV, Load, CurrentGeneration, GetMSSA())) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/LoopUnroll.cpp:316]: Could delay getting MSSA until after the generation check (if that makes a difference...)

**Review by nikic — COMMENTED** [review]
The general approach here seems okay to me. The late unroll pass is awkward because it's at the very end of the pipeline, and it's not really feasible to do much optimization after it. The targeted approach here looks fine in that sense, and doesn't seem to add that much additional code complexity. It's not something that will scale though -- there are more and more optimization we could in theory perform after late unrolling, but we can't reasonably keep replicating them in the post-unroll simplification.

Are you able to share what non-benchmark workload was the original motivation for this?

**Review by fhahn — COMMENTED** [review]
> The general approach here seems okay to me. The late unroll pass is awkward because it's at the very end of the pipeline, and it's not really feasible to do much optimization after it. The targeted approach here looks fine in that sense, and doesn't seem to add that much additional code complexity. It's not something that will scale though -- there are more and more optimization we could in theory perform after late unrolling, but we can't reasonably keep replicating them in the post-unroll simplification.
> 

Agreed, if more cleanups would be beneficial, we likely need to get back to the drawing board.

> Are you able to share what non-benchmark workload was the original motivation for this?

Unfortunately the code is proprietary so I can't share the code, but it is an implementation of Deeplabv3 (https://arxiv.org/abs/1706.05587) tuned for AArch64 and used to blur video backgrounds. The regression is coming from loops with vector intrinsics + unroll pragmas.

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 7]
```diff
@@ -209,13 +212,142 @@ static bool isEpilogProfitable(Loop *L) {
...
+  ScopedHashTable<const SCEV *, LoadValue> AvailableLoads;
+  SmallVector<std::unique_ptr<StackNode>> NodesToProcess;
+  DomTreeNode *HeaderD = DT.getNode(L->getHeader());
+  NodesToProcess.emplace_back(new StackNode(AvailableLoads, 0, HeaderD,
+                                            HeaderD->begin(), HeaderD->end()));
+
+  unsigned CurrentGeneration = 0;
+  while (!NodesToProcess.empty()) {
+    // Grab the first item off the stack. Set the current generation, remove
+    // the node from the stack, and process it.
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/LoopUnroll.cpp:286]: We don't actually remove the node from the stack here -- I'd drop this comment entirely.

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 8]
```diff
@@ -209,13 +212,142 @@ static bool isEpilogProfitable(Loop *L) {
...
+    MemorySSA *MSSA = GetMSSA();
+    if (!MSSA)
+      return nullptr;
+    auto *EarlierMA = MSSA->getMemoryAccess(LV.DefI);
+    MemoryAccess *LaterDef = MSSA->getWalker()->getClobberingMemoryAccess(LI);
+    if (!MSSA->dominates(LaterDef, EarlierMA))
+      return nullptr;
+  }
+
+  if (LV.DefI->getType() != LI->getType())
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Moved, thanks!

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 9]
```diff
@@ -209,13 +212,142 @@ static bool isEpilogProfitable(Loop *L) {
...
+Value *getMatchingValue(LoadValue LV, LoadInst *LI, unsigned CurrentGeneration,
+                        function_ref<MemorySSA *()> GetMSSA) {
+  if (!LV.DefI)
+    return nullptr;
+  if (LV.Generation != CurrentGeneration) {
+    MemorySSA *MSSA = GetMSSA();
+    if (!MSSA)
+      return nullptr;
+    auto *EarlierMA = MSSA->getMemoryAccess(LV.DefI);
+    MemoryAccess *LaterDef = MSSA->getWalker()->getClobberingMemoryAccess(LI);
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Last time I checked it was mostly SCEV, but still good to use BAA, updated, thanks!

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 10]
```diff
@@ -209,13 +212,142 @@ static bool isEpilogProfitable(Loop *L) {
...
+        if (!Load || !Load->isSimple()) {
+          if (I.mayWriteToMemory())
+            CurrentGeneration++;
+          continue;
+        }
+
+        const SCEV *PtrSCEV = SE.getSCEV(Load->getPointerOperand());
+        LoadValue LV = AvailableLoads.lookup(PtrSCEV);
+        if (Value *M = getMatchingValue(LV, Load, CurrentGeneration, GetMSSA)) {
+
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Dropped, thanks!

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 11]
```diff
@@ -209,13 +212,142 @@ static bool isEpilogProfitable(Loop *L) {
...
+          }
+        } else {
+          AvailableLoads.insert(PtrSCEV, LoadValue(Load, CurrentGeneration));
+        }
+      }
+      NodeToProcess->childGeneration(CurrentGeneration);
+      NodeToProcess->process();
+    } else if (NodeToProcess->childIter() != NodeToProcess->end()) {
+      // Push the next child onto the stack.
+      DomTreeNode *child = NodeToProcess->nextChild();
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Fixed, thanks!

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 12]
```diff
@@ -209,13 +212,142 @@ static bool isEpilogProfitable(Loop *L) {
...
+  NodesToProcess.emplace_back(new StackNode(AvailableLoads, 0, HeaderD,
+                                            HeaderD->begin(), HeaderD->end()));
+
+  unsigned CurrentGeneration = 0;
+  while (!NodesToProcess.empty()) {
+    // Grab the first item off the stack. Set the current generation, remove
+    // the node from the stack, and process it.
+    StackNode *NodeToProcess = &*NodesToProcess.back();
+
+    // Initialize class members.
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Dropped, thanks!

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp (hunk 7 above)
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Removed, thanks!

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp [hunk 13]
```diff
@@ -209,13 +212,140 @@ static bool isEpilogProfitable(Loop *L) {
...
+        Node(N), ChildIter(Child), EndIter(End) {}
+  // Accessors.
+  unsigned currentGeneration() const { return CurrentGeneration; }
+  unsigned childGeneration() const { return ChildGeneration; }
+  void childGeneration(unsigned generation) { ChildGeneration = generation; }
+  DomTreeNode *node() { return Node; }
+  DomTreeNode::const_iterator childIter() const { return ChildIter; }
+
+  DomTreeNode *nextChild() {
+    DomTreeNode *child = *ChildIter;
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/LoopUnroll.cpp:246]:
```suggestion
    DomTreeNode *Child = *ChildIter;
```

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Utils/LoopUnroll.cpp (hunk 13 above)
Comment (fhahn) [reply llvm/lib/Transforms/Utils/LoopUnroll.cpp]: Fixed, thanks!

**Pull request merged**